src/Data/Fuzzy/Fuzzy.purs

```diff
@@ -0,0 +1,222 @@
+module Data.Fuzzy where
+
+import Prelude
+
+import Data.Array (snoc, unsnoc)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Ord (genericCompare)
+import Data.Generic.Rep.Show (genericShow)
+import Data.StrMap (StrMap, values)
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.Monoid (mempty)
+import Data.Newtype (class Newtype, unwrap)
+import Data.String (Pattern(..), drop, indexOf, lastIndexOf, take, toLower)
+import Data.String.Utils (length, toCharArray, words)
+import Data.Tuple (Tuple(..))
+
+newtype Fuzzy a = Fuzzy
+  { original :: a
+  , result   :: StrMap (Maybe Result)
+  , score    :: Rank
+  }
+
+derive instance genericFuzzy :: Generic (Fuzzy a) _
+derive instance newtypeFuzzy :: Newtype (Fuzzy a) _
+instance eqFuzzy :: Eq a => Eq (Fuzzy a) where eq = genericEq
+instance showFuzzy :: Show a => Show (Fuzzy a) where show = genericShow
+instance ordFuzzy :: Eq a => Ord (Fuzzy a) where
+  compare (Fuzzy { score }) (Fuzzy { score: score' }) = compare score score'
+
+newtype FuzzyStr = FuzzyStr
+  { result :: Result
+  , score  :: Rank
+  }
+
+derive instance genericFuzzyStr :: Generic FuzzyStr _
+derive instance newtypeFuzzyStr :: Newtype FuzzyStr _
+instance eqFuzzyStr :: Eq FuzzyStr where eq = genericEq
+instance showFuzzyStr :: Show FuzzyStr where show = genericShow
+instance ordFuzzyStr :: Ord FuzzyStr where
+  compare (FuzzyStr { score }) (FuzzyStr { score: score' }) = compare score score'
+
+data Depth = Full | Word | Char
+
+derive instance genericDepth :: Generic Depth _
+instance eqDepth :: Eq Depth where eq = genericEq
+instance showDepth :: Show Depth where show = genericShow
+
+data Pos = Start | Prefix | Mid | Suffix | End
+
+derive instance genericPos :: Generic Pos _
+instance eqPos :: Eq Pos where eq = genericEq
+instance showPos :: Show Pos where show = genericShow
+
+data Rank = Rank Int Int Int Int Int Int | None
+
+derive instance genericRank :: Generic Rank _
+instance eqRank :: Eq Rank where eq = genericEq
+instance showRank :: Show Rank where show = genericShow
+instance ordRank :: Ord Rank where compare = genericCompare
+
+instance semiringRank :: Semiring Rank where
+  add None r = r
+  add r None = r
+  add (Rank u v w x y z) (Rank u' v' w' x' y' z') =
+    Rank (u + u') (v + v') (w + w') (x + x') (y + y') (z + z')
+  zero = None
+  mul None _ = None
+  mul _ None = None
+  mul (Rank u v w x y z) (Rank u' v' w' x' y' z') =
+    Rank (u * u') (v * v') (w * w') (x * x') (y * y') (z * z')
+  one = Rank 1 1 1 1 1 1
+
+type MatchStrAcc =
+  { substr :: String
+  , pos    :: Pos
+  , fuzzy  :: FuzzyStr
+  }
+
+type Result = Array (Either String String)
+
+match :: ∀ a. Boolean -> (a -> StrMap String) -> String -> a -> Maybe (Fuzzy a)
```

Comment:
It's worth explaining what the `extract` function is up to here, mostly for helping us remember what we need to supply ("This takes an arbitrary data structure and turns it into a map of string keys to string values, where the string value is what will actually be fuzzy-matched against") -- unless that's incorrect!

Something we can glance at later on and remember what to supply and how it'll be matched against.